### PR TITLE
[DTensor][1/2][ThreadRNGTracker] Embed sharding spec in CUDA RNG state for global-index-aware random generation

### PR DESCRIPTION
**Pull request by mori360**
### Background
The current OffsetBasedRNGTracker in distributed PyTorch only tracks the RNG state's seed and offset. When used with sharded tensors across multiple devices (ranks), each rank generates random numbers based on a non-global, local offset. This mechanism is inaccurate and results in random number generation that is inconsistent with the deterministic output of a single-device operation. This inconsistency breaks reproducibility in distributed training environments.
The pr aims to fix https://github.com/pytorch/pytorch/issues/168058.

### This PR
This PR introduces sharding_spec at CUDA generator for ThreadBasedRNGTracker in the PRs later. Based on `distribution_elementwise_grid_stride_kernel`, we add `distribution_elementwise_grid_stride_kernel_sharded` to utilize sharding_spec to collecting dtensor spec info and getting consistent results among single device or multi devices.
#### philox generator
philox_4x32_10 generator has the random generation based on seed, subsequence and offset.
With subsequence, we can have the map from each tensor element to subsequence so that we can track random generation no mater how may gpus we have.
e.g. for a full tensor with size of `[world_size]`, it has each element on subsequence from id 0 to `world_size-1`. Then if we  shard it to all the devices, with local tensor size of `[1]` on each device. Through they are the first element locally, but we could utlize sharding_spec to map them from 0 to `world_size-1`, similar to full tensor.

#### config at generator
Added virtual interface methods for thread-based RNG support. User could set by `set_use_thread_based_rng` whit default value as False.
If set_use_thread_based_rng is True, we will go though the contexts below, otherwise we just follow the path before this PR so that **general random generation won't be affected.**

#### sharding_spec
Sharding_spec is only available when set_use_thread_based_rng is True.
Sharding_spec includes `local_shape`, `global_offset`, `global_shape`,`global_strides`. For ThreadBasedRNGTracker to track dtensor sharding info so as to map elements in dtensor.
Note that full tensor could have several tensor dims, we have `local_shape`, `global_offset`, `global_shape`,`global_strides` as arrays with the same length of `num_dim`.
For example, for a tensor with size [2,2], with device mesh size(2,2), we have the spec as
local_shape: (1, 1)
global_offset: (0, 0)
global_shape: (2, 2)
global_strides: (2, 1)

#### sharding_spec mapping to subsequence
We generate the mapping from right to left
```
for i as device dim from right to left
  global_idx_at_i = global_offset[i] + idx % local_shape[i];
  idx /= local_shape[i];
  global_entry_linear_idx += global_idx_at_i * global_strides[i];
```

#### expected state size
After adding sharding_spec, state size increases.
Previously, `state_size = seed_size + offset_size`, and will error if it's not equal, since `seed_size == offset_size`, assume `seed_size = s`, we need `state_size=2s`
Now, `state_size = seed_size + offset_size + local_shape_size + global_offset_size + global_shape_size + global_strides_size `
Here we have `local_shape_size == global_offset_size == global_shape_size == global_strides_size == num_dim * seed_size`, thus we change the RNG state size check to `state_size == 2s + 4*num_dim*s`, if tensor_dim=0, it applies to the previous case

#### unit test
We have `test_thread_rng_tracker_sharding_spec` in this pr, to init sharding_spec, and proof the size of sharding_spec and we can have the same random generation by the same state including sharding_spec

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #171930
* __->__ #171928

### CODE REVIEW

**Comment by wconstab** [comment]
> We generate the mapping from right to left

is this mapping influenced by 'shard order'? should we be following that order instead of assuming r2l order?

File: aten/src/ATen/cuda/CUDAGeneratorImpl.cpp [hunk 1]
```diff
@@ -327,14 +359,32 @@ c10::intrusive_ptr<c10::TensorImpl> CUDAGeneratorImpl::get_state() const {
   // The RNG state comprises the seed, and an offset used for Philox.
   constexpr size_t seed_size = sizeof(uint64_t);
   constexpr size_t offset_size = sizeof(int64_t);
-  constexpr size_t total_size = seed_size + offset_size;
+  const size_t local_shape_size = sizeof(uint64_t) * this->num_dims_;
```
Comment (wconstab) [inline aten/src/ATen/cuda/CUDAGeneratorImpl.cpp]: nit: i think tensor_ndim is a clearer name

File: test/distributed/tensor/test_random_ops.py [hunk 2]
```diff
@@ -734,6 +734,120 @@ def compute_rankwise_if_local_tensor(weight_local, rank):
...
+    @with_comms
+    def test_thread_rng_tracker_sharding_spec(self):
+        """
+        Test that RNG state with sharding specification can be saved and restored correctly.
+
+        This test verifies that:
+        1. Sharding spec (local_shape, global_offset, global_shape, global_strides) can be
+           serialized into the RNG state tensor along with seed and offset.
+        2. The serialized state tensor has the correct format and length.
+        3. After restoring the RNG state, distributed random operations produce
```
Comment (wconstab) [inline test/distributed/tensor/test_random_ops.py]: i would expect another test that simulates different shardings give the same result without using DTensor.

For example, you can create a normal torch.uniform tensor of size 32.  Then you can create 2 tensors of size 16 and you can create another 8 tensors of size 2 and you can show that  each of these has the same 32 values.  When you create multiple tensors, you'd have to modify the RNG state for each creation, similar to how the next PR does it for the tracker.

File: aten/src/ATen/cuda/CUDAGeneratorImpl.cpp [hunk 3]
```diff
@@ -332,15 +368,34 @@ c10::intrusive_ptr<c10::TensorImpl> CUDAGeneratorImpl::get_state() const {
...
   auto rng_state = state_tensor.data_ptr<uint8_t>();
   auto current_seed = this->current_seed();
-  auto offset = static_cast<int64_t>(this->philox_offset_per_thread()); // Note that old THCGeneratorState had offset as std::atomic<int64_t>
+  auto offset = static_cast<int64_t>(this->philox_offset_per_thread());
   memcpy(rng_state, &current_seed, seed_size);
   memcpy(rng_state + seed_size, &offset, offset_size);
 
+  if (use_thread_based_rng_) {
+    const size_t dim_size = sizeof(uint64_t) * this->tensor_ndim_;
+    size_t ptr = seed_size + offset_size;
```
Comment (wconstab) [inline aten/src/ATen/cuda/CUDAGeneratorImpl.cpp:424]: nit: name this sharding_spec?

File: aten/src/ATen/cuda/CUDAGeneratorImpl.cpp [hunk 4]
```diff
@@ -353,16 +408,17 @@ c10::intrusive_ptr<c10::TensorImpl> CUDAGeneratorImpl::get_state() const {
...
   constexpr size_t seed_size = sizeof(uint64_t);
   constexpr size_t offset_size = sizeof(int64_t);
-  constexpr size_t total_size = seed_size + offset_size;
 
   detail::check_rng_state(new_state);
 
   bool no_philox_seed = false;
   auto new_state_size = new_state.numel();
-  if (new_state_size == total_size - offset_size) {
+  if (new_state_size % (4 * seed_size) == seed_size) {
```
Comment (wconstab) [inline aten/src/ATen/cuda/CUDAGeneratorImpl.cpp]: confused about this logic.  maybe this is clever but i feel like it might be better to be explicit about the cases

if (new_state_size == seed_size) {
} else if (new_state_size == seed_size + offset_size) {
} else {
  TORCH_CHECK(new_size - (seed_size + offset_size) % sharding_spec_size == 0)
}

ok i realized that is not correct since we are mixing 2 things in the same check
1) determine if seed or seed + offset is being passed
2) determine if the sharding spec is well formed and extract tensor_ndim

maybe the code would be easier to read if you broke down the 2 steps like that.  First just do logic to figure out 'base_size' based on seed/offset.  Then do logic to work on tensor dim.

File: aten/src/ATen/cuda/CUDAGeneratorImpl.cpp [hunk 5]
```diff
@@ -374,6 +430,41 @@ void CUDAGeneratorImpl::set_state(const c10::TensorImpl& new_state) {
     memcpy(&philox_offset, new_rng_state + seed_size, offset_size);
   }
   this->set_philox_offset_per_thread(static_cast<uint64_t>(philox_offset));
+
+  // Only process sharding_spec if extra data is present in the state
+  size_t base_size = no_philox_seed ? seed_size : seed_size + offset_size;
+  if (new_state_size > static_cast<int64_t>(base_size)) {
+    uint64_t tensor_ndim = (new_state_size - base_size) / (4 * seed_size);
```
Comment (wconstab) [inline aten/src/ATen/cuda/CUDAGeneratorImpl.cpp]: i think i'd rather have named variables for the 4*seed_size (it should be 'sharding_spec_size' right?)

File: aten/src/ATen/cuda/CUDAGeneratorImpl.h [hunk 6]
```diff
@@ -122,6 +123,13 @@ struct CUDAGeneratorState : public c10::intrusive_ptr_target {
   c10::intrusive_ptr<CUDAGeneratorState> clone();
 };
 
+// aten/src/ATen/cuda/detail/OffsetCalculator.cuh
+#if defined(USE_ROCM)
+constexpr int MAX_DIMS = 16;
```
Comment (wconstab) [inline aten/src/ATen/cuda/CUDAGeneratorImpl.h:130]: can you add a comment about this? i guess 16 or 25 is "fine" but its curious to have these specific limits

File: aten/src/ATen/cuda/CUDAGeneratorImpl.h [hunk 7]
```diff
@@ -167,7 +186,24 @@ struct TORCH_CUDA_CPP_API CUDAGeneratorImpl : public c10::GeneratorImpl {
   CUDAGeneratorImpl* clone_impl() const override;
 
   c10::intrusive_ptr<CUDAGeneratorState> state_;
+  uint64_t tensor_ndim_ = 0;
+  std::array<uint64_t, MAX_DIMS> local_shape_{};
```
Comment (wconstab) [inline aten/src/ATen/cuda/CUDAGeneratorImpl.h]: how come we need to allocate these buffers separately from state_ and they need to hold MAX_DIM?

File: aten/src/ATen/cuda/CUDAGeneratorImpl.h [hunk 8]
```diff
@@ -167,7 +186,24 @@ struct TORCH_CUDA_CPP_API CUDAGeneratorImpl : public c10::GeneratorImpl {
...
+  std::array<uint64_t, MAX_DIMS> global_offset_{};
+  std::array<uint64_t, MAX_DIMS> global_shape_{};
+  std::array<uint64_t, MAX_DIMS> global_strides_{};
   std::atomic_flag no_reset_rnn_state_;
+  bool use_thread_based_rng_ = false;
+
+ public:
+  // Returns true if ThreadBasedRNGTracker is being used (sharded kernel),
+  // false for OffsetBasedRNGTracker (original kernel).
```
Comment (wconstab) [inline aten/src/ATen/cuda/CUDAGeneratorImpl.h]: nit: false is also for non-distributed code. Also, this kernel could be used with other non-DTensor systems as well suggest:

```
  // Returns true if cuda RNG is configured for sharding-aware RNG mode,
  // false otherwise
```

File: aten/src/ATen/native/cuda/DistributionTemplates.h [hunk 9]
```diff
@@ -87,6 +88,54 @@ __global__ void distribution_elementwise_grid_stride_kernel(int64_t numel,
...
+          auto [virtual_idx, virtual_offset, single_thread_n] = virtual_idx_func(li);
+          virtual_offset += offset;
+          curand_init(seed, virtual_idx, 4 * (virtual_offset / 4), &state);
+          auto rand = dist_func(&state);
+          transform_func(li, static_cast<accscalar_t>((&rand.x)[virtual_offset % unroll_factor]));
+        }
+      }
+      __syncthreads();
+    }
+  } else {
```
Comment (wconstab) [inline aten/src/ATen/native/cuda/DistributionTemplates.h]: how come we need this else block? it seems like we should only call this kernel if we have a sharding spec, right?

File: aten/src/ATen/native/cuda/DistributionTemplates.h [hunk 10]
```diff
@@ -138,32 +195,94 @@ void distribution_nullary_kernel(at::TensorIteratorBase& iter,
 
   char* out_data = (char*)iter.data_ptr(0);
 
+  uint64_t global_numel = numel;
+  uint64_t single_thread_n = grid.x * block.x;
+  bool has_valid_sharding_spec = false;
+  if (tensor_ndim > 0) {
+    has_valid_sharding_spec = true;
```
Comment (wconstab) [inline aten/src/ATen/native/cuda/DistributionTemplates.h]: i dont really like that we are doing sharding_spec validation in multiple places. we do a TORCH_CHECK in set_sharding_spec which is not exactly the same as this code.  Maybe we should just add a helper function that can do some common things
- validate the spec
- compute the tensor_ndim and return
then we can use this helper here and in set_state?

File: aten/src/ATen/native/cuda/DistributionTemplates.h [hunk 11]
```diff
@@ -87,6 +88,54 @@ __global__ void distribution_elementwise_grid_stride_kernel(int64_t numel,
   }
 }
 
+// grid stride loop kernel for sharded distributions (ThreadBasedRNGTracker)
+// This kernel computes virtual global indices based on sharding_spec to ensure
+// reproducible random number generation across sharded tensors.
```
Comment (wconstab) [inline aten/src/ATen/native/cuda/DistributionTemplates.h]: this comment should explain more details about the design of the kernel

File: aten/src/ATen/native/cuda/DistributionTemplates.h [hunk 12]
```diff
@@ -138,32 +195,94 @@ void distribution_nullary_kernel(at::TensorIteratorBase& iter,
...
+      if (local_shape[i] == 0)
+        has_valid_sharding_spec = false;
+    }
+    auto single_exec_policy = calc_execution_policy(global_numel, unroll_factor);
+    single_thread_n = std::get<1>(single_exec_policy).x * std::get<2>(single_exec_policy).x;
+  }
+  TORCH_CHECK(single_thread_n > 0, "single_thread_n is 0!!!");
+
+  auto virtual_idx_func = [=] __device__(uint64_t local_entry_linear_idx) {
+    if (tensor_ndim == 0) { // not a dtensor
```
Comment (wconstab) [inline aten/src/ATen/native/cuda/DistributionTemplates.h]: similar question as in the kernel- do we care about the non-sharded path for this kernel?

File: aten/src/ATen/native/cuda/DistributionTemplates.h [hunk 13]
```diff
@@ -138,32 +195,94 @@ void distribution_nullary_kernel(at::TensorIteratorBase& iter,
...
+      global_numel *= global_shape[i];
+      if (local_shape[i] == 0)
+        has_valid_sharding_spec = false;
+    }
+    auto single_exec_policy = calc_execution_policy(global_numel, unroll_factor);
+    single_thread_n = std::get<1>(single_exec_policy).x * std::get<2>(single_exec_policy).x;
+  }
+  TORCH_CHECK(single_thread_n > 0, "single_thread_n is 0!!!");
+
+  auto virtual_idx_func = [=] __device__(uint64_t local_entry_linear_idx) {
```
Comment (wconstab) [inline aten/src/ATen/native/cuda/DistributionTemplates.h]: need doc for this function. 
- what the local_entry_linear_idx refers to
- what it returns
- what it's used for

File: torch/distributed/tensor/_random.py [hunk 14]
```diff
@@ -19,10 +19,31 @@
...
+    """Sets whether to use ThreadBasedRNGTracker or OffsetBasedRNGTracker for random ops."""
+    global _USE_THREAD_RNG_TRACKER
+    _USE_THREAD_RNG_TRACKER = use_thread_based
+
+    # If device is provided and is CUDA, also set the flag on the generator
+    if device is not None and device.type == "cuda":
+        device_handle = _get_device_handle(device.type)
+        if device_handle is not None:
+            for generator in device_handle.default_generators:
+                generator.set_use_thread_based_rng(use_thread_based)
```
Comment (wconstab) [inline torch/distributed/tensor/_random.py]: as i keep reading this, i think 'thread_based' is a bad name. it doesn't tell users anything - we aren't using more threads, or something. we should name it descriptively.  
- shard_aware_rng?

File: aten/src/ATen/cuda/CUDAGeneratorImpl.cpp (hunk 4 above)
Comment (mori360) [reply aten/src/ATen/cuda/CUDAGeneratorImpl.cpp]: Yeah, it's a mixture of size check with/without sharding spec.
The code before change should also satisfy the condition check here as a special case with assumption:
1. offset_size = seed_size, total_size = offset_size+seed_size = 2*offset_size
2. sharding_spec_size = 4*offset_size
